### PR TITLE
Issue 1687 Add remove image option for image fields in forms.

### LINKED ISSUE #1687: Add remove image option for image fields in forms.
**Is your feature request related to a problem? Please describe.**

Currently we can't _unset_ an image for any image field in a form.

**Describe the solution you'd like**

Add a _remove image_ button to the image field template to unset images.

**Describe alternatives you've considered**

N/A

**Additional context**

### PR DESCRIPTION
**Pull request by hudson-newey**
# Add remove image option for image fields in forms.

What is the purpose of this PR?
Adds the ability for clients to delete an image (and use the default) for image inputs

## Changes

* Created button on image input field component
* Fixed corrupted images in dev environment

## Problems
There is currently no user feedback that their image has been deleted, or feedback showing the current image.

## Issues

Fixes #1687 

## Visual Changes
remove button

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/components/shared/formly/image-input.component.ts [hunk 1]
```diff
@@ -27,6 +27,12 @@ import { asFormControl } from "./helper";
           [formlyAttributes]="field"
           (ngModelChange)="readFile()"
         />
+
+        <button
+          type="button"
+          (click)="this.formControl.setValue('')"
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.ts]: empty string is not the same as `null`. Use `null` - which means explicitly empty.

Also make this a function in the component rather than an inline callback.

File: src/app/directives/image/image.directive.ts [hunk 2]
```diff
@@ -158,18 +158,19 @@ export class AuthenticatedImageDirective implements OnChanges {
    * Returns true if all other image options are exhausted
    */
   private use404Image(): boolean {
-    return this.images.count() === 0 && this.currentImage === this.defaultImage;
+    return (this.images.count() === 0 && this.currentImage === this.defaultImage);
```
Comment (atruskie) [inline src/app/directives/image/image.directive.ts]: redundant parentheses

File: src/app/directives/image/image.directive.ts [hunk 3]
```diff
@@ -158,18 +158,19 @@ export class AuthenticatedImageDirective implements OnChanges {
...
   /**
    * Returns true if the default image is the only option left available
    */
   private useDefaultImage(): boolean {
-    return (
+    const useDefaultImage = (
       this.images.count() === 0 &&
-      this.currentImage !== this.defaultImage &&
-      !!this.defaultImage
+      this.currentImage !== this.defaultImage
```
Comment (atruskie) [inline src/app/directives/image/image.directive.ts]: explain this change. The previous code seems more correct. Why did you change it?

File: src/app/services/baw-api/baw-api.service.ts [hunk 4]
```diff
@@ -284,13 +284,22 @@ export class BawApiService<
...
     );
 
+    if (model["image"] === "") {
+      const removeImageReqBody = `{ "${model.kind.toLowerCase()}": { "image": null } }`;
+      return request.pipe(
+        mergeMap(() => this.httpPatch(path, removeImageReqBody)),
+        map(this.handleSingleResponse(classBuilder))
+      );
+    }
+
```
Comment (atruskie) [inline src/app/services/baw-api/baw-api.service.ts]: OK; multiple things wrong with this:

- this change is in the wrong spot (it should be withing `hasFormDataOnlyAttributes` and `hasFormDataOnlyAttributes`)
- you shouldn't be using empty string to denote and empty value. `"" !== null`.
- this will only work for fields named `image`
   - the true bug here is with how we serialize fields that must be sent for form-data requests.

**Review by atruskie — CHANGES_REQUESTED** [review]
It looks like you forgot to initialize git-lfs. You'll need to do that and then ensure those BLOBs are removed from the git history.

This change misses the mark. There are two things that need to be done:

- set the image value in the model to `null` with a button (you have almost this)
- serialize the `null` value 
   - you have not done this
   - your attempt to do this is done in the wrong spot.

You'll also need unit tests for both of your changes.

File: src/app/directives/image/image.directive.ts (hunk 3 above)
Comment (hudson-newey) [reply src/app/directives/image/image.directive.ts]: This is an artifact of my testing. It has since been removed

File: src/app/components/shared/formly/image-input.component.ts (hunk 1 above)
Comment (hudson-newey) [reply src/app/components/shared/formly/image-input.component.ts]: This has now been fixed

File: src/app/directives/image/image.directive.ts (hunk 2 above)
Comment (hudson-newey) [reply src/app/directives/image/image.directive.ts]: I've reverted this change

File: src/app/models/AbstractModel.ts [hunk 5]
```diff
@@ -126,7 +133,7 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
        */
       const modelName = snakeCase(this.kind);
       const snakeCaseAttr = snakeCase(attr);
-      output.append(`${modelName}[${snakeCaseAttr}]`, data[attr]);
+      output.append(`${modelName}[${snakeCaseAttr}]`, data[attr] ?? "");
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]: revert

File: src/app/models/AbstractModel.ts [hunk 6]
```diff
@@ -82,7 +82,15 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
...
    */
   public getJsonAttributes(opts?: ModelSerializationOptions): Partial<this> {
-    return this.toObject(this.getModelAttributes(opts), opts);
+    const keys = this.getModelAttributes(opts);
+    const data = this.toObject(keys, opts);
+
+    // if there is an image object attached to the request, remove it
+    if (!isInstantiated(data["image"])) {
+      delete data["image"];
+    }
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]:
```suggestion
    // if there is an image object attached to the request, remove it
    if (!isInstantiated(data["image"])) {
      delete data["image"];
    }
```

this is the wrong place for this change.

also special casing on stringly typed attribute is bad style 

also your comment explains what the code is doing, and not why.

**Review by atruskie — CHANGES_REQUESTED** [review]

File: src/app/models/AbstractModel.ts [hunk 7]
```diff
@@ -207,6 +206,7 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
     if (opts?.create || opts?.update) {
       return this.getPersistentAttributes()
         .filter((meta) => (opts.create ? meta.create : meta.update))
+        .filter((meta) => this[meta.key] instanceof(File) ? opts.formData : true)
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts:213]: Add a comment explaining why this change was made!

**Review by atruskie — CHANGES_REQUESTED** [review]
Ok it looks like you didn't modify the Site model?

Now you just need to add unit tests for the changes.

Much closer!

**Review by atruskie — CHANGES_REQUESTED** [review]
expecting more tests:

- a test that an image component has a remove button
- a test that the image component remove button sets the model value to null
- a test that the image component remove button resets the input field
- a test for each (3 tests) of the serialization functions that the returned models are correct for each of the two states (`null` of `File`)

File: src/app/components/shared/formly/image-input.component.spec.ts [hunk 8]
```diff
@@ -1 +1,78 @@
...
+            expect(getInput()).toBeTruthy();
+        });
+    });
+
+    describe("remove image", () => {
+        it("should set model value to null on click", () => {
+            setup();
+            getButton().click();
+            expect(formGroup.get("asFormControl").value).toBeNull();
+        });
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.spec.ts]: [10:15] Anthony Truskinger
if you were testing an attribute on model is set to null then i'd expect something along the lines of

expect(model.image).toBeNull();

File: src/app/components/shared/formly/image-input.component.spec.ts [hunk 9]
```diff
@@ -1 +1,78 @@
...
+
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.spec.ts:97]: missing test that tests the associated input is cleared when input has a file set

File: src/app/components/shared/formly/image-input.component.ts [hunk 10]
```diff
@@ -59,5 +62,6 @@ export class ImageInputComponent extends FieldType {
   }
   public removeImage(): void {
     this.formControl.setValue(null);
+    this.inputVar.nativeElement.value = "";
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.ts]: why not set this to `null`

File: src/app/models/AbstractModel.spec.ts [hunk 11]
```diff
@@ -182,6 +182,28 @@ describe("AbstractModel", () => {
...
+          class Model extends MockModel {
+            @bawPersistAttr({ supportedFormats: ["formData"] })
+            public value0: any;
+            @bawPersistAttr({ supportedFormats: ["formData"] })
+            public value1: any;
+          }
+          const model = new Model({value0: testFile, value1: "text input"});
+          expect(hasFormDataOnlyAttributes(model)).toBeTrue();
+        });
+
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]: - you didn't test the null state
- you didn't test the other two json serialization functions

**Review by atruskie — CHANGES_REQUESTED** [review]

File: src/app/components/shared/formly/image-input.component.spec.ts (hunk 9 above)
Comment (hudson-newey) [reply src/app/components/shared/formly/image-input.component.spec.ts:97]: I've got this working using a DataTransfer

File: src/app/models/AbstractModel.spec.ts (hunk 11 above)
Comment (hudson-newey) [reply src/app/models/AbstractModel.spec.ts]: The null state check was added in the previous commit (c4dd9597ae1c1c02aa9efa7fb2b4272a084260da) on lines 175-184:
 ```
       it("should return false if only attribute is null", () => {
          class Model extends MockModel {
            @bawPersistAttr({ supportedFormats: ["formData"] })
            public value0: any;
          }
          const model = new Model({value0: null});
          expect(hasFormDataOnlyAttributes(model)).toBeFalse();
        });
```

File: src/app/components/shared/formly/image-input.component.spec.ts [hunk 12]
```diff
@@ -1 +1,92 @@
...
+    it("should set model value to null on click", () => {
+      setup();
+      getButton().click();
+      expect(model.image).toBeNull();
+    });
+
+    it("should remove file from image input field", () => {
+      const dataTransfer = new DataTransfer();
+      const myFile = new File([""], "fileName.png");
+      dataTransfer.items.add(myFile);
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.spec.ts]: explain the usage of data transfer api - or better yet use a mock file

File: src/app/components/shared/formly/image-input.component.spec.ts [hunk 13]
```diff
@@ -1 +1,92 @@
...
+
+      setup();
+      getInput().files = dataTransfer.files;
+      getButton().click();
+      expect(getInput().value).toBeFalsy();
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.spec.ts]: cache your `getInput` call

Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.spec.ts]: your arrange and assertion change different properties. Are you sure this works?

I'd expect this: 
```
getInput().value = file;
// act
 expect(getInput().value).toBeFalsy();
```

or 

```
      getInput().files = dataTransfer.files;
      expect(getInput().value).toBeTruthy();
      getButton().click();
      expect(getInput().value).toBeFalsy();
```

File: src/app/components/shared/formly/image-input.component.ts [hunk 14]
```diff
@@ -20,18 +20,27 @@ import { asFormControl } from "./helper";
...
+          type="button"
+          (click)="removeImage()"
+          class="btn btn-outline-danger"
+        >Remove</button>
       </div>
     </div>
   `,
 })
 export class ImageInputComponent extends FieldType {
+  @ViewChild("imageInput", {static: false})
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.ts]: is the `static` property needed? if not remove, if so explain why

File: src/app/components/shared/formly/image-input.component.ts [hunk 15]
```diff
@@ -20,18 +20,27 @@ import { asFormControl } from "./helper";
...
+  public inputVar: ElementRef;
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.ts]: bad name, suggestion:
```suggestion
  public imageInput: ElementRef;
```

File: src/app/components/shared/formly/image-input.component.ts [hunk 16]
```diff
@@ -51,4 +60,9 @@ export class ImageInputComponent extends FieldType {
 
     this.formControl.setValue(images.item(0));
   }
+  public removeImage(): void {
+    this.formControl.setValue(null);
+    this.model.image = null;
+    this.inputVar.nativeElement.value = null;
```
Comment (atruskie) [inline src/app/components/shared/formly/image-input.component.ts]: something seems fishy here - we're updating too many things. 

try removing the `setValue` or `model.image = null` lines to see if they're actually needed

File: src/app/models/AbstractModel.spec.ts [hunk 17]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
     });
   });
 
+  describe("nullEmit", () => {
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]:
```suggestion
  describe("emitting file or null in formdata or json requests", () => {
```

File: src/app/models/AbstractModel.spec.ts [hunk 18]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
...
+    function hasFormDataOnlyAttributes(model: AbstractModel) {
+      return model.hasFormDataOnlyAttributes({ create: true });
+    }
+    const testFile = new File([""], "testFileName.png");
+
+    it("should emit null type in JSON attributes", () => {
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]:
```suggestion
    it("should emit null when serializing a model and deleting a file", () => {
```

File: src/app/models/AbstractModel.spec.ts [hunk 19]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
...
+      class Model extends MockModel {
+        @bawPersistAttr({ create: true, supportedFormats: ["json", "formData"] })
+        public image: any;
+      }
+      const model = new Model({ id: 1, image: null });
+      expect(Object.keys(model.getJsonAttributes({ create: true }))).toContain("image");
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]: This test could emit any value other than `null` and it wouldn't fail.

```suggestion
      var actual = model.getJsonAttributes({ create: true });
      expect(actual).toHaveOwnProperty("image");  // not sure if this API exists
      expect(actual.image).toBe(null);
      // or 
      expect(actual).toEqual(jasmine.objectContaining({
        image: null
      }));
```

Comment (atruskie) [reply src/app/models/AbstractModel.spec.ts]: hasOwnProperty

File: src/app/models/AbstractModel.spec.ts [hunk 20]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
...
+    it("should not emit file type object in JSON attributes", () => {
+      class Model extends MockModel {
+        @bawPersistAttr({ create: true, supportedFormats: ["json", "formData"] })
+        public image: any;
+      }
+      const model = new Model({ id: 1, image: testFile });
+      expect(Object.keys(model.getJsonAttributes({ create: true }))).not.toContain("image");
+    });
+
+    it("should return false if attribute is null", () => {
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]:
```suggestion
    it("should not want to send a formdata request if attribute is null", () => {
```

File: src/app/models/AbstractModel.spec.ts [hunk 21]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
...
+    it("should return true if attribute is a File type object", () => {
+      class Model extends MockModel {
+        @bawPersistAttr({ supportedFormats: ["json", "formData"] })
+        public image: any;
+      }
+      const model = new Model({image: testFile});
+      expect(hasFormDataOnlyAttributes(model)).toBeTrue();
+    });
+
+    it("should convert null type values to JSON", () => {
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]:
```suggestion
    it("toJSON emits values as is (null case)", () => {
```

File: src/app/models/AbstractModel.spec.ts [hunk 22]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
...
+      expect(model.toJSON()).toBeNull();
+    });
+
+    it("should not convert image type values to JSON", () => {
+      class Model extends MockModel {
+        @bawPersistAttr({ create: true, supportedFormats: ["json", "formData"] })
+        public image: any;
+      }
+      const model = new Model({ id: 1, image: testFile });
+      expect(typeof model.toJSON() === "object").toBeTrue();
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts]: This is an invalid assertion.

```suggestion
      var actual = model.toJSON();
      expect(actual.image).toBeInstanceOf(File);
```

File: src/app/models/AbstractModel.spec.ts [hunk 23]
```diff
@@ -132,6 +132,67 @@ describe("AbstractModel", () => {
...
+    });
+  });
+
```
Comment (atruskie) [inline src/app/models/AbstractModel.spec.ts:242]: you're missing tests for the formdata serialization method `getFormDataOnlyAttributes`

File: src/app/models/AbstractModel.ts [hunk 24]
```diff
@@ -73,8 +73,11 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
     return this.viewUrl;
   }
 
-  public toJSON(): Partial<this> {
-    return this.toObject(this.getModelAttributes());
+  public toJSON(): null | File {
+    const emptyFile: File = new File([""], "");
+    const jsonObject = Object.values(this.toObject(this.getModelAttributes()));
+
+    return jsonObject.includes(Object(emptyFile)) ? emptyFile : null;
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]: revert  - this is incorrect

File: src/app/models/AbstractModel.ts [hunk 25]
```diff
@@ -207,6 +209,8 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
     if (opts?.create || opts?.update) {
       return this.getPersistentAttributes()
         .filter((meta) => (opts.create ? meta.create : meta.update))
+        // The following filter forces null values from formData (e.g. from the "Remove" image button) to send via JSON
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]:
```suggestion
        // The following filter splits values for attributes that support both json and formData formats
        // when a  null value is present, we send the value in the json request
        // when a File value is present, we send the value in the formData request
        // The null/json scenario is used to support deleting images.
```

**Review by atruskie — CHANGES_REQUESTED** [review]

**Review by atruskie — APPROVED** [review]
Ok let's squash your changes and then we'll merge
